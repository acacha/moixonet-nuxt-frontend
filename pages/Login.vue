<template>
  <v-card class="elevation-12">
    <v-toolbar
      color="primary"
      dark
      flat
    >
      <v-toolbar-title data-test="auth_login_card_title">
        Si us plau, entreu...
      </v-toolbar-title>
    </v-toolbar>
    <v-card-text>
      <v-form>
        <v-text-field
          v-model="email"
          data-test="auth_login_input_email"
          label="Correu electrÃ²nic"
          name="login"
          prepend-icon="mdi-email"
          type="text"
        />

        <v-text-field
          id="password"
          v-model="password"
          data-test="auth_login_input_password"
          label="Paraula de pas"
          name="password"
          prepend-icon="mdi-lock"
          type="password"
        />
      </v-form>
    </v-card-text>
    <v-card-actions>
      <nuxt-link to="/register">
        Registrar-se
      </nuxt-link>&nbsp;<nuxt-link to="/password/reset">
        Recordar paraula de pas
      </nuxt-link>
      <v-spacer />
      <v-btn :loading="loading" @click="login" color="primary" data-test="auth_login_button_login">
        Entrar
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script>
export default {
  name: 'Login',
  layout: 'auth',
  data () {
    return {
      email: '',
      password: '',
      loading: false
    }
  },
  methods: {
    login () {
      this.loading = true
      this.$auth.loginWith('local', {
        data: {
          email: this.email,
          password: this.password
        }
      }).then(() => {
        this.loading = false
      })
    }
  }
}
</script>

<style scoped>

</style>
