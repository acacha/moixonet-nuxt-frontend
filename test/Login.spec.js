import { shallowMount, mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vue from 'vue'
import flushPromises from 'flush-promises'
import Login from '../pages/Login'

Vue.use(Vuetify)
describe('Login', () => {
  let opts
  const auth = {
    loginWith: jest.fn().mockImplementation(() => Promise.resolve())
    // ALTERNATIVES
    // loginWith: jest.fn().mockResolvedValue(43)
    // loginWith: jest.fn()
    // loginWith: jest.fn().mockRejectedValue('Error occurred')
    // loginWith: jest.fn().mockImplementation(() => Promise.reject('Error occurred'));
  }
  const build = (options) => {
    options = options || opts
    const wrapperShallow = shallowMount(Login, options)
    const wrapper = mount(Login, options)

    return {
      wrapperShallow,
      loginButton: () => wrapper.find('[data-test=auth_login_button_login]'),
      emailInputField: () => wrapper.find('[data-test=auth_login_input_email]'),
      passwordInputField: () => wrapper.find('[data-test=auth_login_input_password]')
    }
  }

  beforeEach(() => {
    opts = {
      stubs: ['nuxt-link'],
      sync: false,
      mocks: {
        $auth: auth
      },
      vuetify: new Vuetify({
        mocks: {
          $vuetify: {
            breakpoint: {} // STUB BREAKPOINTS SERVICE: https://vuetifyjs.com/en/getting-started/unit-testing#mocking-vuetify
          }
        }
      })
    }
  })

  test('is a Vue instance', () => {
    const { wrapperShallow } = build()
    expect(wrapperShallow.isVueInstance()).toBeTruthy()
  })

  test('renders the component', () => {
    const { wrapperShallow } = build()
    expect(wrapperShallow.element).toMatchSnapshot()
  })

  test('can login', () => {
    const { loginButton, emailInputField, passwordInputField } = build()
    emailInputField().element.value = 'sergiturbadenas@gmail.com'
    emailInputField().trigger('input')
    passwordInputField().element.value = '12345678'
    passwordInputField().trigger('input')
    loginButton().trigger('click')

    expect(auth.loginWith).toHaveBeenCalled()
    expect(auth.loginWith.mock.calls[0][1].data).toEqual({ email: 'sergiturbadenas@gmail.com', password: '12345678' })
  })

  test('shows error on error login', async () => {
    const auth = {
      loginWith: jest.fn().mockRejectedValue('Error occurred')
    }
    const snackbar = {
      showError: jest.fn()
    }
    opts = {
      stubs: ['nuxt-link'],
      sync: false,
      mocks: {
        $auth: auth,
        $snackbar: snackbar
      },
      vuetify: new Vuetify({
        mocks: {
          $vuetify: {
            breakpoint: {} // STUB BREAKPOINTS SERVICE: https://vuetifyjs.com/en/getting-started/unit-testing#mocking-vuetify
          }
        }
      })
    }
    const { loginButton, emailInputField, passwordInputField } = build(opts)
    emailInputField().element.value = 'sergiturbadenas@gmail.com'
    emailInputField().trigger('input')
    passwordInputField().element.value = '12345678'
    passwordInputField().trigger('input')
    loginButton().trigger('click')

    await flushPromises()

    expect(snackbar.showError).toHaveBeenCalled()
    expect(snackbar.showError.mock.calls[0][0]).toEqual('Error occurred')
  })
})
