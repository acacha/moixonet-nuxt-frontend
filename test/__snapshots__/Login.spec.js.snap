// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the component 1`] = `
<v-card-stub
  class="elevation-12"
  loaderheight="4"
  tag="div"
>
  <v-toolbar-stub
    color="primary"
    dark="true"
    extensionheight="48"
    flat="true"
    src=""
    tag="header"
    tile="true"
  >
    <v-toolbar-title-stub
      data-test="auth_login_card_title"
      datatest="auth_login_card_title"
    >
      
      Si us plau, entreu...
    
    </v-toolbar-title-stub>
  </v-toolbar-stub>
   
  <v-card-text-stub>
    <v-form-stub>
      <v-text-field-stub
        backgroundcolor=""
        clearicon="$clear"
        data-test="auth_login_input_email"
        errorcount="1"
        errormessages=""
        label="Correu electrònic"
        loaderheight="2"
        messages=""
        name="login"
        prependicon="mdi-email"
        rules=""
        successmessages=""
        type="text"
        value=""
      />
       
      <v-text-field-stub
        backgroundcolor=""
        clearicon="$clear"
        data-test="auth_login_input_password"
        errorcount="1"
        errormessages=""
        id="password"
        label="Paraula de pas"
        loaderheight="2"
        messages=""
        name="password"
        prependicon="mdi-lock"
        rules=""
        successmessages=""
        type="password"
        value=""
      />
    </v-form-stub>
  </v-card-text-stub>
   
  <v-card-actions-stub>
    <nuxt-link-stub
      to="/register"
    >
      
      Registrar-se
    
    </nuxt-link-stub>
     
    <nuxt-link-stub
      to="/password/reset"
    >
      
      Recordar paraula de pas
    
    </nuxt-link-stub>
     
    <v-spacer-stub />
     
    <v-btn-stub
      activeclass=""
      color="primary"
      data-test="auth_login_button_login"
      tag="button"
      type="button"
    >
      
      Entrar
    
    </v-btn-stub>
  </v-card-actions-stub>
</v-card-stub>
`;
